Create Proposed Constitution (#2)
freechipsproject/chisel-lang-governance: constitution.md

designed to support the needs of a variety of RTL development communities, including industry, academia, and individuals.

It includes support for the “Chisel stack”, including but not limited to FIRRTL, Testers, Chisel Utilities, documentation, 
and administrative work.

## Members
The Chisel self-governance consists of five types of members:
**Technical Advisory Committee members**, **administrators**, **reviewers**, **developers**,
and **volunteers**.

### Technical Advisory Committee
The role of Technical Advisory Committee (TAC) is to represent the interests of the
projects and members of the Chisel Working Group. This includes, but is not limited
to, facilitating communication and collaboration between members, as well as interfacing
with CHIPS Alliance on behalf of the Chisel community. Additionally, the TAC is
responsible for upholding the policies outlined in this document.

Other responsibilities include: admitting (and if necessary expelling) projects,
organizing community meetings/conferences, resolving disputes within and between projects,
allocating funds, clarifying procedures for each membership type, and publicizing
the Chisel community.

> Projects are admitted (or expelled) by a 2/3 vote of the TAC.

Reviewer: ```suggestion
Projects are admitted (or expelled) by a 2/3 vote of the TAC, assuming a quorum is present.
```

A list of current projects is maintained in a document in this repository.

The TAC is made up of members of the community as well as one member appointed by
the CHIPS Alliance Board.
The members are listed by name and Github username in a document within
this repository. Membership changes are reflected by pull requests
against the membership list document. Members have the power to remove themselves.
Membership additions (or expulsions) require 2/3 approval of the current members
of the TAC.

#### TAC Chair
The chair of the TAC is selected from and by the TAC members by popular vote.
The chair's responsibilities are to schedule and run meetings of the TAC, as well
as count votes and enforce voting periods for any action that requires a vote.

The chair will serve until they step down or are replaced by a new vote of the TAC.
Chair elections generally occur during scheduled meetings, but in the event of an
absent or otherwise misbehaving chair, such a vote can be initiated by any TAC member
as a PR against the TAC membership document.

> 

Reviewer: ```suggestion

#### Voting Procedure
Voting can only proceed if a quorum of 2/3 of all TAC members is established. An abstained vote counts towards quorum but not towards determining whether a motion passes or fails. Voting rules and procedure is enforced by the TAC Chair.
```

this repository. Membership changes are reflected by pull requests
against the membership list document. Members have the power to remove themselves.
Membership additions (or expulsions) require 2/3 approval of the current members
of the TAC.

#### TAC Chair
The chair of the TAC is selected from and by the TAC members by popular vote.
The chair's responsibilities are to schedule and run meetings of the TAC, as well
as count votes and enforce voting periods for any action that requires a vote.

The chair will serve until they step down or are replaced by a new vote of the TAC.
Chair elections generally occur during scheduled meetings, but in the event of an
absent or otherwise misbehaving chair, such a vote can be initiated by any TAC member
as a PR against the TAC membership document.

#### Voting Procedure
Voting can only proceed if a quorum of 2/3 of all TAC members is established. An abstained vote counts towards quorum but not towards determining whether a motion passes or fails. Voting rules and procedure is enforced by the TAC Chair.
### Administrator
Consists of one lead and any number of additional administrators who direct a

> given project and report progress to the TAC

Reviewer 2: Missing period at end of sentence

> given project and report progress to the TAC

Reviewer: ```suggestion
given project and report progress to the TAC.
```

the CHIPS Alliance Board.

## Mission Statement
The Chisel ecosystem is a stable, user-friendly platform for production-quality RTL generator development, 
designed to support the needs of a variety of RTL development communities, including industry, academia, and individuals.

It includes support for the “Chisel stack”, including but not limited to FIRRTL, Testers, Chisel Utilities, documentation, 
and administrative work.

## Members
The Chisel self-governance consists of five types of members:
**Technical Advisory Committee members**, **administrators**, **reviewers**, **developers**,
and **volunteers**.

### Technical Advisory Committee
The role of Technical Advisory Committee (TAC) is to represent the interests of the
projects and members of the Chisel Working Group. This includes, but is not limited
to, facilitating communication and collaboration between members, as well as interfacing
with CHIPS Alliance on behalf of the Chisel community. Additionally, the TAC is
responsible for upholding the policies outlined in this document.

Other responsibilities include: admitting (and if necessary expelling) projects,

> organizing community meetings/conferences, resolving disputes between projects,

Reviewer: ```suggestion
organizing community meetings/conferences, resolving disputes within and between projects,
```

### Developers
Developers are the most common membership type, and is an entry point for enthusiastic community members. They have voting
rights for high-level directives as set by the administrators of a given project.

Anyone can become a developer by signing up and *signing the CLA*. In exchange for voting rights, a developer must commit to 
5 hours per month of development work on any Chisel project. This work can include:
* creating pull requests
* reviewing pull requests
* submitting issues
* discussing issues
* answering Stack Overflow questions
* attending meetings
* external contributions (registered on website via submissions)
  * writing blogs
  * additional documentation
* red lighting/green lighting releases

Developers who have been inactive for six months or more will be removed from the group.

A developer interested in becoming a reviewer should have a wide body of commits and begin reviewing submitted pull 
requests to practice and build up a body of reviews, both to demonstrate their expertise but also to increase their 
familiarity with the repository. These pull-requests will also require reviews from a Reviewer. Once a developer has 

> had over 10 reviews on a project, they can submit a request to the administrators to become a Reviewer for that project.

Reviewer: ```suggestion
had over 10 reviews on a project, the individual can nominate themselves, or the community can nominate them, to become a Reviewer. A request is then sent to the administrators, who decide whether to add the individual as a Reviewer for that project.
```

Developers can contribute feedback by testing scheduled releases and fixes, and these members are critical to ensuring 
Chisel's backwards compatibility, as well as discovering bugs before a release. 

### Volunteers
Volunteers are people who want to contribute to a Chisel project, but do not commit to a regular contribution. They do 
not have voting rights, and no registration is required. Volunteers may attend development meetings. Volunteers may 

> contribute through PR’s to Chisel repositories, and are encouraged to review and comment but they cannot accept PR requests. It is hoped that volunteers will become developers.

Reviewer: ```suggestion
contribute through PRs to Chisel repositories, and are encouraged to review and comment but they cannot accept PR requests. It is hoped that volunteers will become developers.
```

## Collaboration Policy

> The CWG’s standard communication channels and development policies are outlined as follows, and will likely be adopted for other 

Reviewer: ```suggestion
The CWGs standard communication channels and development policies are outlined as follows, and will likely be adopted for other 
```

associated CHIPS Alliance working groups.

## Communication Channels
High level leadership directives will be published via mailing list, and minutes of these discussions will be posted on the 
Chisel website.

All technical development discussions will be done either on Github with Github issues, pull requests, and comments, or via 
a recurring open development meeting with video conferencing. Open development meetings will be run by project 
administrators or senior reviewers.

User-facing support is primarily through answering questions on **Stack Overflow**, contributing to and maintaining a **Github 
Wiki page**, and promoting community dialogue with an associated **Gitter** channels. The purpose of Gitter is to move discussion 
from mailing lists to a more modern communication outlet.

> 

Reviewer: ```suggestion
User-facing support is primarily through a commitment to answering questions, providing documentation, and promoting community dialogue via a variety of communication channels listed on the website.

```

### Development Policy
Generally, mature projects will release minor/major versions, where minor versions are backwards compatible, 
and major versions can break old APIs. Old APIs will be deprecated prior to removal.

> New features which are experimental could be released in an **experimental** package, whereby all API’s in this package are 

Reviewer: ```suggestion
New features which are experimental could be released in an **experimental** package, whereby all APIs in this package are 
```

contributor for other behaviors that they deem inappropriate, threatening, offensive, or harmful.

### Scope
This Code of Conduct applies both within project spaces and in public spaces when an individual is representing the project 
or its community. Examples of representing a project or community include using an official project e-mail address, posting 
via an official social media account, or acting as an appointed representative at an online or offline event. 
Representation of a project may be further defined and clarified by project administrators.

### Enforcement
Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team leader. 
All complaints will be reviewed and investigated and will result in a response that is deemed necessary and appropriate to 
the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an incident. 
Further details of specific enforcement policies may be posted separately.

Project administrators who do not follow or enforce the Code of Conduct in good faith may face temporary or permanent 

> repercussions as determined by other members of the project's leadership.

Reviewer: ```suggestion

All Contributors are expected to abide by the code of conduct detailed in `CODE_OF_CONDUCT.md`, which is a verbatim copy of [Contributor Covenant 2.0](https://www.contributor-covenant.org/version/2/0/code_of_conduct/code_of_conduct.md).

```

Reviewer 3: I think similar concerns apply to using this verbatim though in this case.

# Chisel Working Group (CWG) Provisional Constitution

### Table of Contents
* [Relationship to CHIPS Alliance](#relationship-to-chips-alliance)
* [Mission Statement](#mission-statement)
* [Members](#members)
  * [Technical Advisory Committee](#technical-advisory-committee)
  * [Administrator](#administrator)
  * [Reviewers](#reviewers)
  * [Developers](#developers)
  * [Volunteers](#volunteers)
* [Collaboration Policy](#collaboration-policy)
  * [Communication Channels](#communication-channels)
  * [Development Policy](#development-policy)
* [Amending the Constitution](#amending-the-constitution)
* [TAC Voting](#tac-voting)
* [Code of Conduct (Contributors)](#code-of-conduct-contributors)
  * [Our Pledge](#our-pledge)
  * [Our Standards](#our-standards)
  * [Our Responsibilities](#our-responsibilities)
  * [Scope](#scope)
  * [Enforcement](#enforcement)
  
## Relationship to CHIPS Alliance

The Chisel Working Group is a member of CHIPS Alliance and subject to the bylaws established by
the CHIPS Alliance Board.

## Mission Statement

> The Chisel ecosystem is a stable, user-friendly platform for production-quality RTL generator development, 

Reviewer 3: ```suggestion
The Chisel ecosystem is intended to provide a stable, user-friendly platform for production-quality RTL generator development, 
```

Reviewer 4: I don't think this language needs to be weakened.

Reviewer: Is this satisfactory to leave, @edwardcwang ?

Reviewer 3: My personal take is that this comes across a little bit as self-congratulatory - e.g. "mission accomplished, we are the best". Though it's also fine with me to leave it as-is.

designed to support the needs of a variety of RTL development communities, including industry, academia, and individuals.

It includes support for the “Chisel stack”, including but not limited to FIRRTL, Testers, Chisel Utilities, documentation, 
and administrative work.

## Members
The Chisel self-governance consists of five types of members:
**Technical Advisory Committee members**, **administrators**, **reviewers**, **developers**,
and **volunteers**.

### Technical Advisory Committee
The role of Technical Advisory Committee (TAC) is to represent the interests of the
projects and members of the Chisel Working Group. This includes, but is not limited
to, facilitating communication and collaboration between members, as well as interfacing
with CHIPS Alliance on behalf of the Chisel community. Additionally, the TAC is
responsible for upholding the policies outlined in this document.

Other responsibilities include: admitting (and if necessary expelling) projects,

> organizing community meetings/conferences, resolving disputes between projects,

Reviewer 2: perhaps a nit, but also resolving disputes within projects, were the community can't come to a consensus

allocating funds, clarifying procedures for each membership type, and publicizing

> the Chisel community.

Reviewer 4: Should delegation be explicitly called out as a responsibility? For example, it could be the TAC's role to delegate some of it's responsibilities to a sub-committee they select. I suppose delegation is more of an ability than it is a responsibility, but it seems like something that would be good to mention somewhere.

Reviewer: I think delegation is implicit in any leadership position.

> Projects are admitted (or expelled) by a 2/3 vote of the TAC.

Reviewer 2: Clarification: 2/3 of all TAC members (so abstain counts as a no vote), or 2/3 of TAC members who cast a vote (so abstaining doesn't affect the vote proportion either way)? Or more broadly, is abstaining allowed, and is that the default for anyone who doesn't vote?

Reviewer: @seldridge will do some research and plan is to discuss at next meeting.

A list of current projects is maintained in a document in this repository.

The TAC is made up of members of the community as well as one member appointed by
the CHIPS Alliance Board.
The members are listed by name and Github username in a document within

> this repository. Membership change is reflected by pull requests

Reviewer 3: ```suggestion
this repository. Membership changes are reflected by pull requests
```

enforcing collaboration policies, and determining how contributors join the project.
They also resolve disagreements between Reviewers, Developers, Volunteers, and users.

Administrators have all of the rights and responsibilities of Reviewers.

### Reviewers
Reviewers are selected per project repository, and their primary responsibility is to maintain the health of the code 
repositories by enforcing a consistent coding style and good software engineering practices. Their second responsibility 
is to encourage community growth by mentoring enthusiastic contributors.

*Reviewers will be the members who most directly interact with the developer community, so it is of primary importance that 
they encourage a welcoming atmosphere.*

Reviewers are the only members who have write-privileges to their repository, so they must be active regularly such that 
development will not stall on their reviews. The procedure for becoming a Reviewer is outlined in the following Developers 
section.

Reviewers strictly subsume Developers (e.g. they also have all the rights and responsibilities of Developers).

### Developers

> Developers are the most common membership type, and is an entry point for enthusiastic community members. They have voting

Reviewer 3: Has this ever happened / has this happened yet?

Reviewer: No, primarily because we haven't had a constitution yet. Do you have an objection to this idea?

rights for high-level directives as set by the administrators of a given project.

> Anyone can become a developer by signing up and *signing the CLA*. In exchange for voting rights, a developer must commit to 

Reviewer 3: Especially given this

5 hours per month of development work on any Chisel project. This work can include:
* creating pull requests
* reviewing pull requests
* submitting issues
* discussing issues
* answering Stack Overflow questions
* attending meetings
* external contributions (registered on website via submissions)
  * writing blogs
  * additional documentation
* red lighting/green lighting releases

Developers who have been inactive for six months or more will be removed from the group.

A developer interested in becoming a reviewer should have a wide body of commits and begin reviewing submitted pull 
requests to practice and build up a body of reviews, both to demonstrate their expertise but also to increase their 
familiarity with the repository. These pull-requests will also require reviews from a Reviewer. Once a developer has 

> had over 10 reviews on a project, they can submit a request to the administrators to become a Reviewer for that project.

Reviewer 2: I think it should also be allowed (perhaps even encouraged) for the community to nominate new Reviewers (in addition to being able to self-nominate).

Developers can contribute feedback by testing scheduled releases and fixes, and these members are critical to ensuring 
Chisel's backwards compatibility, as well as discovering bugs before a release. 

### Volunteers
Volunteers are people who want to contribute to a Chisel project, but do not commit to a regular contribution. They do 
not have voting rights, and no registration is required. Volunteers may attend development meetings. Volunteers may 

> contribute through PR’s to Chisel repositories, and are encouraged to review and comment but they cannot accept PR requests. It is hoped that volunteers will become developers.

Reviewer 4: Very minor nit, but I think it is incorrect to use an apostrophe with an initialism. `s/PR's/PRs/`. Other examples below.

## Collaboration Policy

> CWG’s standard communication channels and development policy is outlined as follows, and will likely be adopted for other 

Reviewer 3: ```suggestion
The CWG’s standard communication channels and development policies are outlined as follows, and will likely be adopted for other 
```

associated CHIPS Alliance working groups.

## Communication Channels
High level leadership directives will be published via mailing list, and minutes of these discussions will be posted on the 
Chisel website.

All technical development discussions will be done either on Github with Github issues, pull requests, and comments, or via 
a recurring open development meeting with video conferencing. Open development meetings will be run by project 
administrators or senior reviewers.

User-facing support is primarily through answering questions on **Stack Overflow**, contributing to and maintaining a **Github 

> Wiki page**, and promoting community dialogue with an associated **Gitter** channels. The purpose of Gitter is to move discussion 

Reviewer 3: Should we really hardcode Gitter into the constitution? I was thinking almost leaving it as a chat room / channel (e.g. especially with our experiment with matrix etc)

Reviewer 5: Probably not. This is likely worth revising.

Reviewer: Agreed. Weird to have this in the constitution. Let's remove all explicit communication app mentions and instead have 'a commitment to communication by a variety of means`?

from mailing lists to a more modern communication outlet.

### Development Policy

> Generally, development on mature products will release minor/major versions, where minor versions are backwards compatible, 

Reviewer 3: ```suggestion
Generally, mature projects will release minor/major versions, where minor versions are backwards compatible, 
```

> Generally, development on mature products will release minor/major versions, where minor versions are backwards compatible, 

Reviewer 4: Should we explicitly reference semantic versioning?

Reviewer: Ok to be high level in the constitution, as changes to versioning likely shouldn't require a change to the constitution.

> and major versions can break ok API’s. Old API’s will be deprecated prior to removal.

Reviewer 3: ```suggestion
and major versions can break old APIs. Old APIs will be deprecated prior to removal.
```

I assume ok = old?

New features which are experimental could be released in an **experimental** package, whereby all API’s in this package are 
subject to API changes without deprecation, and may change on a minor release (although this will be avoided if possible). 
Successful experimental features/APIs will eventually be migrated to other packages.

All code contributions to mature projects (Chisel, FIRRTL, etc.) must undergo a review process and pass continuous 

> integration tests prior to merging into the master branch.

Reviewer 3: While I am a 100% CI superfan, sometimes CI does introduce issues (gets hung up, some yosys check fails for silly reasons, etc). Maybe we can add a clause (if it's not already elsewhere) to add a backup provision for this?

Reviewer 5: Maybe just "pass testing". We already have the ability to merge PRs that rightly fail tests (`[skip chisel tests]` in FIRRTL). Also, this does technically enshrine `master` branch into the constitution which implies a specific development model. Maybe we can generalize this?

I think, broadly, this is trying to say that:

1. We have a culture of testing
1. Projects should have reviewers
1. People should not be committing directly to a release branch
1. These rules *may* not apply if you have a very immature project

Reviewer: Replace with 'pass continuous integration tests' with 'rigorous testing' ?

> integration tests prior to merging into the master branch.

Reviewer: ```suggestion
All code contributions to mature projects (Chisel, FIRRTL, etc.) must undergo a review process and pass rigorous testing prior to merging into the master branch.
```

Changes to the constitution are proposed as pull requests against this document.
They are approved by 2/3 vote of the TAC as expressed by TAC members approving
the pull request on Github.

## TAC Voting
For events that require a vote of the TAC (eg. changing the constitution,
admitting new projects, admitting new members, selecting or changing TAC chair,
expelling projects, expelling members), votes may occur via TAC members approving
(or not approving) Github pull requests against this repository. Voting periods for
a given change are set by the chair to a length of time not less than 2 weeks.
Votes are tallied and if approved, pull requests are merged by the TAC Chair.

Because many decisions may be non-controversial or involve personnel decisions better
discussed in a meeting than via pull requests, votes **excluding changing the
constitution** may occur in official TAC meetings. These votes will be tallied by the
TAC chair, and if approved, reflected as pull requests by the chair against this
respository. The chair must mention in the pull request when the vote took place, and
another member of the TAC must approve the pull request for it to be merged.

> ## Code of Conduct (Contributors)

Reviewer 3: Possibility to add a clause indicating that the overall goal of this CoC is to advance the technical development of the Chisel ecosystem. CoCs have been controversial in some projects where it has derailed technical development

Reviewer: After some discussion, let's just take an existing code of conduct, which has been vetted, verbatim.

Adapted from https://github.com/YosysHQ/yosys, which is itself adapted from the Contributor Covenant, version 1.4, 
available at http://contributor-covenant.org/version/1/4/

### Our Pledge

> In the interest of fostering an open and welcoming environment, we as contributors and administrators pledge to make 

Reviewer 3: To avoid issues where X group feels like they are not included in the special protected list from Y-ism this could be shortened to

```
In the interest of fostering an open and welcoming environment, we as contributors and administrators pledge to make participation in our project and our community a harassment-free experience for everyone.
```

> participation in our project and our community a harassment-free experience for everyone, regardless of age, bodysize, 

Reviewer 4: I haven't seen bodysize as one word before. Should it be two?

> participation in our project and our community a harassment-free experience for everyone, regardless of age, bodysize, 

Reviewer: ```suggestion
participation in our project and our community a harassment-free experience for everyone, regardless of age, body size, 
```
